### ISSUE TITLE
 Validate SBOM for .NET Repos using Arcade main

### ISSUE THREAD

**Issue by epananth** [dotnet/arcade#8477]
•	**If your repo is using Arcade from the ‘.NET Eng – latest’ channel and using Arcade’s ([jobs.yml]) template to build**, you should just need the latest arcade update to get SBOM generation automatically added to your pipelines.
•	If your repo is not using Arcade’s templates, or not using Arcade at all, you will need to manually add the SBOM generation task manually to every build job that creates or modifies assets. You can follow the steps outlined [here](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Farcade%2Fblob%2Fmain%2FDocumentation%2FSBOMGenerationGuidance.md%23repositories-not-using-arcades-jobsyml-templates&data=04%7C01%7CEpsitha.Ananth%40microsoft.com%7C892581b8ac764a9b9e4a08d9f2588d46%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637807284437427727%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=OyqzofPWBXZBorUNFRNpxv0jhJe5%2BWmuw0UAGANMeHE%3D&reserved=0) to use a helper template that we’re providing through Arcade.
 
•	**Action required by 2/25/2022- SBOM validation for repos using Arcade main**: We need to make sure all repositories are generating SBOMs as part of their official builds, and that those SBOMs meet certain initial requirements. Follow the steps outlined [here](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Farcade%2Fblob%2Fmain%2FDocumentation%2FSBOMGenerationGuidance.md%23reviewing-generated-sboms-for-correctness&data=04%7C01%7CEpsitha.Ananth%40microsoft.com%7C892581b8ac764a9b9e4a08d9f2588d46%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637807284437427727%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=UPQOokU5K7KdGfTPp%2FlFsGpASkXu2No5Dp%2BiEf5vpkI%3D&reserved=0) to validate the generated SBOMs, and update status below when you have completed the work. Note that if two people are editing the issue, one of the changes might get lost, so double check that your information is recorded appropriately.

•	For repositories that produce assets released via the .NET release pipeline or if your repo name is in the list [here](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Farcade%2Fblob%2Fmain%2FDocumentation%2FSBOMGenerationGuidance.md%23retention-rules-for-release-build-sboms&data=04%7C01%7CEpsitha.Ananth%40microsoft.com%7Cffd8a1d65c6a42d53ccd08d9f249f388%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637807221711878073%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=lOvlw7x3bDxlIFNLVRrRdDbJigFyrIljN45sDbQnDQE%3D&reserved=0), your builds are automatically retained.
•	For repositories that have their own release process, you can follow the steps outlined [here](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Farcade%2Fblob%2Fmain%2FDocumentation%2FSBOMGenerationGuidance.md%23retention-rules-for-release-build-sboms&data=04%7C01%7CEpsitha.Ananth%40microsoft.com%7Cffd8a1d65c6a42d53ccd08d9f249f388%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637807221711878073%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=lOvlw7x3bDxlIFNLVRrRdDbJigFyrIljN45sDbQnDQE%3D&reserved=0)

| Status         | Description |
|--------------  | ----------- |
|     ✔️| Results verified. Good to go! |
|   ❌  | Did not work | 

| Repository |  Owner | Status | Does this need sbom? | Notes |
|-------------|---------|---------|------|--------|
| Nuget.BuildTasks |  @MiYanni  |  |  | Spl case, gets arcade update and inserts only to VS | 
| ASP.Net Classic nuget packages  |  @StephenMolloy  |  |  | New case, has not been updated in a few years (added 3/8 to this list) | 
| aspnet-AspNetWebHooks |  @dougbu | ❌ | N/A | we haven't built this in ages | 
| aspnet-AspNetWebStack |  @dougbu | ❌ | | we do not have a pipeline to build this; currently builds on TeamCity | 
| AzureSignalR-samples | @Y-Sindo |  ❌| N/A | No pipeline |
| aspnet-Benchmarks | @sebastienros |  ❌| N/A| No pipeline | 
| aspnet-SignalR-Client-Cpp | @BrennanConroy |  ❌| N/A| Ships as code |
| dotnet-project-system |  @MiYanni | ✔️ | Spl case, inserts to VS |
| dotnet-project-system-tools |  @MiYanni | ✔️| | Update done, need to verify |
| microsoft-dotnet-framework-docker | @mthalman |  ❌ | N/A | N/A - only produces Windows Docker images, SBOM only supports Linux Docker images | 
| microsoft-go | @dagood| ❌ | | Not part of .NET |
| microsoft-go-images | @dagood| ❌ | | Not part of .NET | 
| microsoft-go-infra | @dagood  | ❌ | N/A | No pipeline: source is the asset. Not part of .NET |
| microsoft-go-infra-images | @dagood| ❌| | Not part of .NET | 
| dotnet-diagnostictests | @hoyosjs   | ❌| | Test only repo does not require SBOM generation|
| SignalR-SignalR | @BrennanConroy | ✔️ | |  Working on arcade udpate | 
| dotnet-insertions-client |  @bekir-ozturk| | || 
| dotnet-roslyn-tools |  @JoeRobich |✔️ | |https://github.com/dotnet/roslyn-tools/pull/1171 |
| dotnet-source-indexer | @alexperovich | ❌ | N/A | Is a website  | 
| dotnet-arcade-extensions | @riarenas  |✔️ | | | 
| dotnet-aspnetcore | @dougbu | | | https://github.com/microsoft/dropvalidator/issues/397 | 
| dotnet-crank | @sebastienros  |✔️| Uses arcade release/6.0  | Created on 2/21 | 
| dotnet-deployment-tools | @NikolaMilosavljevic  | ✔️ | | working on it  | 
| dotnet-diagnostics-internal-components |   @hoyosjs| | |Look for successful build | 
| dotnet-efcore | @dougbu | ✔️| | has manifests for jobs that never publish and list _everything_ in the artifacts/ folder | 
| dotnet-emsdk |  @lewing/@akoeplinger |✔️| | | 
| dotnet-fsharp | @brettfo/@kevinransom  | ✔️|| Needs arcade update | 
| dotnet-interactive | @colombod  |✔️ | |Needs arcade update | 
| dotnet-interactive-window | @tmat  |✔️| |  | 
| dotnet-iot | @joperezr  | ❌| N/A |Spl repo uses arcade but is not microsoft owned | 
| dotnet-llvm-project | @akoeplinger  | | |(https://github.com/microsoft/dropvalidator/issues/368)  | 
| dotnet-machinelearning-assets | @ericstj |✔️ | | arcade update is flowing? | 
| dotnet-machinelearning| @ericstj  |✔️ |  | |
| dotnet-maui | @antonfirsov  | | |arcade update is flowing? | 
| dotnet-microsoft.maui.graphics | @mjbond-msft  | || arcade update is flowing? | 
| dotnet-msbuild| @rainersigwald   | ✔️| | Needs arcade update | 
| dotnet-optimization | @DrewScoggins | | | Need work | 
| dotnet-razor-compiler | @dougbu | ✔️| | | 
| dotnet-razor-tooling | @NTaylorMullen | ✔️| | Needs arcade update | 
| dotnet-roslyn |  @JoeRobich |✔️ | |Still needs update to reference SBOM from VS Component manifests | 
| dotnet-roslyn-debug |  @tmat |  | |arcade update is flowing? | 
| dotnet-roslyn-sdk |  @JoeRobich |✔️ | |https://github.com/dotnet/roslyn-sdk/pull/970 | 
| dotnet-sdk | @marcpopMSFT  | ✔️ | | | 
| dotnet-source-build | @MichaelSimons | ❌ | N/A | No source/build in main  | 
| dotnet-source-build-reference-packages | @MichaelSimons | ❌|  N/A | Does not produce shippable packages | 
| dotnet-source-build-utilities | @MichaelSimons | ❌ | N/A | Internal non-shipping source-build tooling  | 
| dotnet-sourcelink |  @tmat |✔️| || 
| dotnet-spark | @suhsteve  | | |Needs arcade update | 
| dotnet-symuploader | @hoyosjs  | | |Not sure | 
| dotnet-templating | @vlada-shubina  | ✔️ | |Needs arcade update | 
| dotnet-test-templates | @Haplois   | ✔️| |Needs arcade update | 
| dotnet-try-convert | @jmarolf  |✔️ | |last update was in oct 21st| 
| dotnet-upgrade-assistant | @sunandabalu  |✔️ | |arcade update taken on 2/16 | 
| dotnet-wcf | @HongGit  | ✔️| |Needs arcade update | 
| dotnet-windowsdesktop | @dreddy-work  |✔️ | |Needs update | 
| dotnet-winforms-designer | @Shyam-Gupta  |✔️ | |Verified that SBOM is getting generated correctly | 
| dotnet-winforms | |✔️ | | | 
| dotnet-winforms-datavisualization | @RussKie  | | |last update was in dec 31st| 
| dotnet-wpf-int | @singhashish-wpf | | |Needs arcade update | 
| Microsoft-clrmd | @leculver  | | |Needs arcade update | 
| dotnet-docker-tools | @mthalman  | | N/A |N/A - Doesn't produce anything shipped to customers |
| microsoft-vstest | @Evangelink  | |  ||
| Nuget.Client | @zivkan  |✔️ |  ||
| vs-code-coverage |  | |  ||
| dotnet-arcade |  @epananth| ✔️|  | |
| dotnet-arcade-services | @epananth | ✔️|  | | 
| dotnet-arcade-validation | @epananth | ✔️| | | 
| aspnet-AspLabs | @dougbu | ✔️ | | has manifests for jobs that never publish and list _everything_ in the artifacts/ folder |
| aspnet-AspNetKatana | @Tratcher  | | | Repo does not use Arcade; builds on TeamCity and will move to Azdo in future, they are working on this and will add sbom in future  | 
| dotnet-ef6 | @dougbu  | ✔️| | hasn't released in ages and may never again /cc @ajcvickers; oddly configured, does not create a MergedManifest.xml |
| dotnet-runtime-assets |  @lewing |✔️| | | 
| dotnet-command-line-api|  @vlada-shubina | ✔️| | | 
| dotnet-diagnostics|  @hoyosjs | ✔️|  ||
| dotnet-dotnet-monitor| @jander-msft  | ✔️|  ||
| dotnet-helix-machines| @epananth | ✔️| | |
| dotnet-helix-service| @epananth | ✔️|  | |
| dotnet-hotreload-utils| @akoeplinger  |✔️ |  | |
| dotnet-HttpRepl   | @tlmii  |✔️|  | |
| dotnet-icu |  @lewing |✔️| | |
| dotnet-installer | @marcpopMSFT  | ✔️ | |  |
| dotnet-linker | @sbomer  |✔️ |  | |
| dotnet-metadata-tools | @tmat |✔️ | |  |
| dotnet-msquic | @wfurt | ✔️|  | |
| dotnet-performance | @wfurt | ✔️|  | |
| dotnet-release | @epananth | ✔️ |  | |
| dotnet-roslyn-analyzers | @JoeRobich |✔️ | |  |
| dotnet-runtime | @agocke  |  ✔️ || https://github.com/microsoft/dropvalidator/issues/397 |
| dotnet-Scaffolding | @deepchoudhery | ✔️ | |  |
| dotnet-source-build-externals | @MichaelSimons   |  ✔️ | | |
| dotnet-symreader | @tmat  |✔️ |  ||
| dotnet-symreader-converter | @tmat |✔️  | |  |
| dotnet-symreader-portable| @tmat  |✔️ ||  |
| dotnet-symstore|  @hoyosjs  |✔️ |  ||
| dotnet-tye|  @philliphoff | ✔️ |  ||
| dotnet-winforms-designer | @dreddy-work | ✔️| |  |
| dotnet-wpf | @singhashish-wpf  | ✔️|  ||
| dotnet-xharness |  @epananth | ✔️ |  ||
| dotnet-xliff-tasks |  @epananth| ✔️|  ||
| microsoft-reverse-proxy |  @MihaZupan | ✔️ | | |
| dotnet-cli-lab |  @joeloff | ✔️   | | Internal build with changes succeeded, waiting to merge PR |

**Comment by dougbu** [comment]
I would appreciate more context here

1. How urgent are the "Needs arcade update" items❔ ASP.NET-related repos tend to take Arcade updates ~weekly.
2. What is the point of creating an SBOM in a repo like aspnet/AspLabs which ships one package on a very irregular basis❔ We're moving that source from AspLabs to aspnetcore in any case.
3. I don't know much about aspnet-AspNetKatana @Tratcher is this yours❔
4. [aspnet-AspNetWebStack](https://dev.azure.com/dnceng/internal/_git/aspnet-AspNetWebStack/branches) is a special case. We have an internal repo for security fixes but no AzDO pipeline. Plan is to migrate builds from TeamCity to AzDO before the end of the year. Not sure how we'd create an SBOM using the current infrastructure. Suggestions❔
5. @ajcvickers should you own dotnet-ef6❔ Or is it time to declare we'll never do another EF6 release and kill that repo and associated pipelines❔

**Comment by mmitche** [comment]
> I would appreciate more context here
> 
> 1. How urgent are the "Needs arcade update" items❔ ASP.NET-related repos tend to take Arcade updates ~weekly.
> 2. What is the point of creating an SBOM in a repo like aspnet/AspLabs which ships one package on a very irregular basis❔ We're moving that source from AspLabs to aspnetcore in any case.
> 3. I don't know much about aspnet-AspNetKatana @Tratcher is this yours❔
> 4. [aspnet-AspNetWebStack](https://dev.azure.com/dnceng/internal/_git/aspnet-AspNetWebStack/branches) is a special case. We have an internal repo for security fixes but no AzDO pipeline. Plan is to migrate builds from TeamCity to AzDO before the end of the year. Not sure how we'd create an SBOM using the current infrastructure. Suggestions❔
> 5. @ajcvickers should you own dotnet-ef6❔ Or is it time to declare we'll never do another EF6 release and kill that repo and associated pipelines❔

1. We'd like to have SBOMs generated in main by end of next week. You probably already got the update in main though (it got checked in froday
2. Microsoft wants SBOMs for all of its supply chain, even for repos that are only providing artifacts for other repos. Since asplabs ships to customers, it should have an SBOM
3. ...
4. This one can be skipped then.
5. Up to you folks there.

**Comment by dougbu** [comment]
> Since asplabs ships to customers, it should have an SBOM

Agree that's the current state. It however shouldn't ship anything in the future, given the move to dotnet/aspnetcore. Right @dotnet/aspnet-build @JamesNK❔

**Comment by JamesNK** [comment]
The gRPC source is moving from asplabs to aspnetcore. It won't ship from there anymore.

However, I think it is valuable to have a place where our product team can have experiments that can be published to NuGet to get feedback.

**Comment by dougbu** [comment]
> However, I think it is valuable to have a place where our product team can have experiments that can be published to NuGet to get feedback.

That makes it a very weird case from an SBOM perspective because the pipeline build more than we'd ever release.

**Comment by mmitche** [comment]
> > However, I think it is valuable to have a place where our product team can have experiments that can be published to NuGet to get feedback.
> 
> That makes it a very weird case from an SBOM perspective because the pipeline build more than we'd ever release.

You could argue that about any of the .NET core repos. There will be thousands of SDK builds that never get released. The SBOM generation itself is pretty cheap, so I don't see a huge issue here

**Comment by lewing** [comment]
I have even less context here, what is needed beyond an arcade update?

**Comment by epananth** [comment]
@Lewing : If your repo is using Aracade's (job.yml/jobs.yml) to build, SBOM generation is already in place for this scenario. So once you get an arcade update (https://github.com/dotnet/arcade/blob/main/Documentation/SBOMGenerationGuidance.md#repositories-using-arcades-jobsyml-templates)

You should be able to review SBOM (please take a look here on how to verify)(https://github.com/dotnet/arcade/blob/main/Documentation/SBOMGenerationGuidance.md#reviewing-generated-sboms-for-correctness)

**Comment by Y-Sindo** [comment]
Regarding to repo AzureSignalR-samples, we use it as a place for samples only and there is no pipeline for the repo.

**Comment by MihaZupan** [comment]
> `Package name and version`: After the packages section, the last entry should mention the correct name and version for the software that the SBOM is about. The name property should read as ".NET 7.0.0" for main branches, and ".NET 6.0.0" for .NET 6 release branches.

If the name expected to be `.NET 7.0.0` even for projects that don't ship as part of .NET (microsoft/reverse-proxy)?

**Comment by sebastienros** [comment]
__dotnet/crank__ has "Has latest arcade update but still does not have sbom generated" comment.
My response to this is it's using arcade release/6.0 apparently, and I saw the changes with SBOM just got pushed there. So I believe dotnet/crank will get it when there is a new release of that branch?

__aspnet/benchmark__ doesn't have a pipeline. We are not shipping anything from this repository, it's just a set of benchmarks applications and scripts.

**Comment by Tratcher** [comment]
> 3\. I don't know much about aspnet-AspNetKatana @Tratcher is this yours❔

Yes, that's mine. That repo doesn't use Arcade yet. We might consider switching to Arcade when we move it from TeamCity to Azdo.

**Comment by vlada-shubina** [comment]
I checked dotnet/templating and SBOM is generated in internal build from main branch. 
The file is very big but I checked that key outputs are present.

**Comment by epananth** [comment]
> Regarding to repo AzureSignalR-samples, we use it as a place for samples only and there is no pipeline for the repo.

I will update the list. thanks for confirming

**Comment by dagood** [comment]
> dotnet-diagnostictests

I don't know this repo, so I've changed the owner from myself to ❓

**Comment by epananth** [comment]
> > `Package name and version`: After the packages section, the last entry should mention the correct name and version for the software that the SBOM is about. The name property should read as ".NET 7.0.0" for main branches, and ".NET 6.0.0" for .NET 6 release branches.
> 
> Is the name expected to be `.NET 7.0.0` even for projects that don't ship as part of .NET (microsoft/reverse-proxy)?

For now all the repos using arcade main to use package name as .NET 7.0.0 

> > `Package name and version`: After the packages section, the last entry should mention the correct name and version for the software that the SBOM is about. The name property should read as ".NET 7.0.0" for main branches, and ".NET 6.0.0" for .NET 6 release branches.
> 
> Is the name expected to be `.NET 7.0.0` even for projects that don't ship as part of .NET (microsoft/reverse-proxy)?

Yes, we want everyone who uses arcade main to use .NET 7.0.0 as package name

**Comment by BrennanConroy** [comment]
How do we handle repos that
a. Don't use .NET at all
b. Don't ship a physical package
Specifically talking about aspnet-SignalR-Client-Cpp which is fully C++ and ships as code (i.e. you compile it yourself).

**Comment by MichaelSimons** [comment]
@epananth - I am having some troubles with dotnet-source-build-reference-packages and am wondering if you can help out.  I got the [repo updated](https://github.com/dotnet/source-build-reference-packages/pull/337) on the latest arcade version but the build is still not running the SBOM generation leg - https://dev.azure.com/dnceng/internal/_build/results?buildId=1619144&view=logs&s=6884a131-87da-5381-61f3-d7acc3b91d76&j=2f0d093c-1064-5c86-fc5b-b7b1eca8e66a.  The build is utilizing the Arcade job template as illustrated [here](https://github.com/dotnet/source-build-reference-packages/blob/main/azure-pipelines/builds/ci.yml#L18)

**Comment by epananth** [comment]
> How do we handle repos that a. Don't use .NET at all b. Don't ship a physical package Specifically talking about aspnet-SignalR-Client-Cpp which is fully C++ and ships as code (i.e. you compile it yourself).

I will have to get back to you on this one.

**Comment by epananth** [comment]
> @epananth - I am having some troubles with dotnet-source-build-reference-packages and am wondering if you can help out. I got the [repo updated](https://github.com/dotnet/source-build-reference-packages/pull/337) on the latest arcade version but the build is still not running the SBOM generation leg - https://dev.azure.com/dnceng/internal/_build/results?buildId=1619144&view=logs&s=6884a131-87da-5381-61f3-d7acc3b91d76&j=2f0d093c-1064-5c86-fc5b-b7b1eca8e66a. The build is utilizing the Arcade job template as illustrated [here](https://github.com/dotnet/source-build-reference-packages/blob/main/azure-pipelines/builds/ci.yml#L18)

Taking a look

**Comment by epananth** [comment]
We don't need to produce sboms for dotnet-source-build-reference-packages. Going to update the list.

**Comment by MiYanni** [comment]
@epananth `dotnet-project-system` doesn't use Arcade.

**Comment by epananth** [comment]
@MiYanni Updated dotnet-project-system, what about dotnet-project-system-tools?

**Comment by zivkan** [comment]
@epananth NuGet.Client does not use Arcade, so I'm working through trying to get it all working.

MicroBuild/DevDiv requires the sbom to be included in the vsix. Taking a quick look at the docs you linked, it looks like dnceng only requires the sbom to be published as a build artifact. Is that correct?  There's no requirement to embed the sbom in nupkgs promoted to BAR or published to nuget.org?

**Comment by epananth** [comment]
> @epananth NuGet.Client does not use Arcade, so I'm working through trying to get it all working.
> 
> MicroBuild/DevDiv requires the sbom to be included in the vsix. Taking a quick look at the docs you linked, it looks like dnceng only requires the sbom to be published as a build artifact. Is that correct? There's no requirement to embed the sbom in nupkgs promoted to BAR or published to nuget.org?

Yes it need to be published as a build artifact, no need to publish to BAR

**Comment by Forgind** [comment]
For dotnet/msbuild:
I tried to follow the steps to verify our generated SBOM is correct. I found one thing I think is a problem and two things I wanted to ask about.
The license information on our packages is "NOASSERTION" instead of MIT license. I think this needs fixing?
The name property of spdxVersion is .NET 6.0.0 rather than .NET 7.0.0 even though this is from main. rainersigwald said that is correct because we're shipping in SDK 6.0.300 still.
I checked that Microsoft.Build.dll, Microsoft.NET.StringTools.dll, Microsoft.Build.Tasks.Core.dll, and Microsoft.Build.Conversion.Core.dll were all present in both the packages section and the files section, though I didn't verify that they were present in all flavors. We have other assemblies, but that's a representative sample. Is there something else I should verify here?

**Comment by BillHiebert** [comment]
@hoyosjs - that repo is still in use as it is where the hot reload components live.

**Comment by epananth** [comment]
> nupkgs

1) We are not concerned about overreporting at this point, so we want to include everything for now. 
2)  All licenses are NOASSERTION, all download locations are NOASSERTION - we do not have an option to change this for now, I am going to talk to SBOM folks and get back on this.
3) Files section does not have nupkgs - can you point the build to me?

I will update the owners in the issue.

**Comment by dougbu** [comment]
Oh, I forgot: The SBOM generation fails on Linux MUSL x64 but produces an empty artifact named after the job. Please either shut it off on that platform or find a way to fix the bug. In the meantime, we can skip CG on that platform; can we skip SBOM generation too❔

**Comment by dougbu** [comment]
Most of my comments above were based on aspnetcore-ci-official. In other news, [AspLabs-ci-official](https://dev.azure.com/dnceng/internal/_build?definitionId=249) and [efcore-ci-official ](https://dev.azure.com/dnceng/internal/_build?definitionId=229) seem to produce SBOM files and artifacts in jobs that list mostly test results and obj/ files in the SBOM files. The jobs also produce packages and those are listed but only the Windows job in either build would ever be used to ship anything. How should I handle this extreme over-reporting?

For the future, can SBOM generation be disabled in jobs that do not produce a manifest XML file❔

**Comment by dougbu** [comment]
To get a reasonable comparison between the SBOM files and asset manifests, I adjusted my script to _exclude_ files named `*-ci*` and _include_ non-shipping files from asset manifests. I suggest all but dotnet-aspnetcore should be checked off but leave that decision to you @epananth.

### Results

- aspnet-AspLabs looks fine though there's a **lot** of useless information in the extra SBOM files
- dotnet-aspnetcore is missing a few shipping assets due to the inability to produce SBOM files on Linux MUSL x64
- dotnet-ef6 looks fine
- dotnet-efcore looks fine though there's a **lot** of useless information in the extra SBOM files
- dotnet-razor-compiler looks fine

#### More details

<details>
<summary>aspnet-AspLabs</summary>

``` text
*asplabs.xml --> manifest.asplabs*.json
  [-] MergedManifest.xml
```
</details>

<details>
<summary>dotnet-aspnetcore</summary>

``` text
*aspnetcore.xml --> manifest.aspnetcore*.json
  [-] aspnetcore-runtime-7.0.0-preview.3.22117.6-linux-musl-x64.tar.gz
  [-] aspnetcore-runtime-7.0.0-preview.3.22117.6-linux-musl-x64.tar.gz.sha512
  [-] aspnetcore-runtime-internal-7.0.0-preview.3.22117.6-linux-musl-x64.tar.gz
  [-] aspnetcore-runtime-internal-7.0.0-preview.3.22117.6-linux-musl-x64.tar.gz.sha512
  [-] MergedManifest.xml
  [-] Microsoft.AspNetCore.App.Runtime.linux-musl-x64.7.0.0-preview.3.22117.6.nupkg
  [-] Microsoft.AspNetCore.App.Runtime.linux-musl-x64.7.0.0-preview.3.22117.6.symbols.nupkg
  [-] Microsoft.SourceBuild.Intermediate.aspnetcore.7.0.0-preview.3.22117.6.nupkg
  [-] Microsoft.SourceBuild.Intermediate.aspnetcore.7.0.0-preview.3.22117.6.symbols.nupkg
```
</details>

<details>
<summary>dotnet-ef6</summary>

``` text
*ef6.xml --> manifest.ef6*.json
```
</details>

<details>
<summary>dotnet-efcore</summary>

``` text
*efcore.xml --> manifest.efcore*.json
  [-] MergedManifest.xml
```
</details>

<details>
<summary>dotnet-razor-compiler</summary>

``` text
*razor-compiler.xml --> manifest.razor-compiler*.json
  [-] MergedManifest.xml
  [-] Microsoft.SourceBuild.Intermediate.razor-compiler.7.0.0-preview.3.22116.4.nupkg
  [-] Microsoft.SourceBuild.Intermediate.razor-compiler.7.0.0-preview.3.22116.4.symbols.nupkg
```
</details>

**Comment by dougbu** [comment]
Not sure what to do about aspnet-AspNetWebHooks. We have a pipeline but it doesn't use Arcade and only publishes to the general-testing and general-testing-symbols feeds. The repo contains legacy (desktop) ASP.NET WebHooks and we last published packages from it to NuGet in 2019. We have no plans for another release.

**Comment by dougbu** [comment]
Are we supposed to check .NET 6 as well❔

**Note**: aspnet-AspLabs publishes to the `.NET 6` channel but uses Arcade from the `.NET Eng - Latest` channel. This confuses the SBOM files. They use the name `.NET 7.0.0`.

**Comment by epananth** [comment]
@dougbu those are great suggestions, I created an issue to look into this in future -> https://github.com/dotnet/core-eng/issues/15671

> * At a minimum, publish the SBOM files to a single artifact with appropriately unique names
>   
>   * This is required even if the files are kept separate (see (2)) because the number of artifacts is unweildy
> * Better to merge build-specific manifest.spdx.json files into a single one, akin to MergedManifest.xml
> * Some type of validation e.g. a comparison w/ a fake merged manifest in PR validation would be ideal

We unfortunately do not want to to turn off sbom generation for now. We have an issue opened for Sbom folks (https://github.com/microsoft/dropvalidator/issues/368) Once that is fixed that should go away.

> Most of my comments above were based on aspnetcore-ci-official. In other news, [AspLabs-ci-official](https://dev.azure.com/dnceng/internal/_build?definitionId=249) and [efcore-ci-official ](https://dev.azure.com/dnceng/internal/_build?definitionId=229) seem to produce SBOM files and artifacts in jobs that list mostly test results and obj/ files in the SBOM files. The jobs also produce packages and those are listed but only the Windows job in either build would ever be used to ship anything. How should I handle this extreme over-reporting?
> 
> For the future, can SBOM generation be disabled in jobs that do not produce a manifest XML file❔

Overreporting at this stage is intentional, so we want to keep it that way for now. 

> 

thanks for confirming the list of repos. I will update the issue. 

> Not sure what to do about aspnet-AspNetWebHooks. We have a pipeline but it doesn't use Arcade and only publishes to the general-testing and general-testing-symbols feeds. The repo contains legacy (desktop) ASP.NET WebHooks and we last published packages from it to NuGet in 2019. We have no plans for another release.

I think we don't have to create SBOM for this, since you have no plans for another release.

> Are we supposed to check .NET 6 as well❔
> 
> **Note**: aspnet-AspLabs publishes to the `.NET 6` channel but uses Arcade from the `.NET Eng - Latest` channel. This confuses the SBOM files. They use the name `.NET 7.0.0`.

Not yet. 

I will get back to you on aspnet-AspLabs

**Comment by dougbu** [comment]
@epananth thanks for your responses.

> We have an issue opened for Sbom folks (https://github.com/microsoft/dropvalidator/issues/368) Once that is fixed that should go away.

I don't believe that issue really covers the problems doing SBOM generation on Linux MUSL x64 machines. Is there another issue to track for that❔

**Comment by epananth** [comment]
I am going to run another build with better verbosity and look into that further. Will keep you posted..

**Comment by mmitche** [comment]
> > 3. I don't know much about aspnet-AspNetKatana @Tratcher is this yours❔
> 
> Yes, that's mine. That repo doesn't use Arcade yet. We might consider switching to Arcade when we move it from TeamCity to Azdo.

@Tratcher We need determine the applicability of the EO to the repo. Can you figure out whether katana is in scope for 3/8 based on the Scope section here? https://eng.ms/docs/initiatives/executive-order/executive-order-requirements/executiveorderoncybersecurity/overview

If it is, then the TeamCity dependency will need to be removed. Is there a timeline on that?

**Comment by Tratcher** [comment]
Katana produces the Microsoft.Owin libraries used to build a variety of cloud services so it probably qualifies as a top level dependency? I don't see anything in the list that would disqualify it.

We're planning to move it off TeamCity in the next few months, but don't have a specific schedule yet.

**Comment by joeloff** [comment]
@mmitche @marcpopMSFT Should we include https://github.com/dotnet/cli-lab? That's where the uninstall tool lives.

**Comment by mmitche** [comment]
> @mmitche @marcpopMSFT Should we include https://github.com/dotnet/cli-lab? That's where the uninstall tool lives.

Yeah, I think so /cc @epananth

**Comment by epananth** [comment]
> > @mmitche @marcpopMSFT Should we include https://github.com/dotnet/cli-lab? That's where the uninstall tool lives.
> 
> Yeah, I think so /cc @epananth

Added to the list.

**Comment by zivkan** [comment]
Update on NuGet.Client:

NuGet.Client is a mono-repo that builds both our VS "extension" and the code that gets inserted into the .NET SDK. I imagine that Roslyn is in a similar position. However, NuGet.Client does not use Arcade or Helix or basically any dnceng infrastructure. For testing our VS extension, we use VSEng's DartLab, which requires us to create a "VS bootstrapper". MicroBuild's "Create VS Bootstrapper" doesn't currently (well, as of Tuesday this week) strip out the sbom from the `vsman` files, which the "real" VS build does, which means that NuGet's build is failing to install VS on DartLab, which means we can't get green builds to merge my change.

Additionally, NuGet.Client's builds have always pre-created/downloaded some test packages that test execution will use to validate the build. These packages are not used to build NuGet, not even build the tests, but 1ES's sbom.json generation detects these packages addinng a LOT of noise into the sbom.json file, including packages that don't have license information. Due to the short notice of this sbom requirement, we haven't had time to redesign our builds to avoid this problem. Our first implementation of sbom in our builds will contain this noise, and we'll schedule work in the future to try to improve our pipeline, but I personally and hoping that 1ES improve their sbom task to do analysis of builds to detect what's actually used, rather than just detect all packages on the CI agent.

TL;DR, I have a branch/PR with minimum changes to comply with both dnceng and VS requirements for sbom, even if the quality of the contents of the sbom.json is poor. I'm blocked on a MicroBuild pipeline task bug before I can get a green build to merge my PR.

**Comment by mmitche** [comment]
> 

This is a good approach. The tooling will improve over time, and likely will eventually become per-package (this is my educated guess) rather than "gather everything up and mash it into one file". Get something minimal working now, and we'll iterate later.

**Comment by jander-msft** [comment]
I cannot edit this issue, so here are the results I'd like to report:
- Repository: dotnet-dotnet-monitor
- Ownership: please change to @jander-msft
- Status: Verified

I'd also like to report that the Linux Musl (Alpine) x64 build leg fails to generate an SBOM and creates an empty artifact. However, we do not ship files out of this build leg. See https://dev.azure.com/dnceng/internal/_build/results?buildId=1625839&view=logs&j=ce9b67a1-188c-57b1-9fb6-8fdc7e08cad8&t=bfa3c9d4-d8b2-5ecb-1e13-ed53d43bfaa5 for the example failure.

**Comment by epananth** [comment]
> I cannot edit this issue, so here are the results I'd like to report:
> 
> * Repository: dotnet-dotnet-monitor
> * Ownership: please change to @jander-msft
> * Status: Verified
> 
> I'd also like to report that the Linux Musl (Alpine) x64 build leg fails to generate an SBOM and creates an empty artifact. However, we do not ship files out of this build leg. See https://dev.azure.com/dnceng/internal/_build/results?buildId=1625839&view=logs&j=ce9b67a1-188c-57b1-9fb6-8fdc7e08cad8&t=bfa3c9d4-d8b2-5ecb-1e13-ed53d43bfaa5 for the example failure.

Update the issue. Thanks for validating @jander-msft . For follow up on linux-musl, I created https://github.com/microsoft/dropvalidator/issues/368. Waiting to hear from SBOM folks.

**Comment by epananth** [comment]
> @epananth thanks for your responses.
> 
> > We have an issue opened for Sbom folks ([microsoft/dropvalidator#368](https://github.com/microsoft/dropvalidator/issues/368)) Once that is fixed that should go away.
> 
> I don't believe that issue really covers the problems doing SBOM generation on Linux MUSL x64 machines. Is there another issue to track for that❔

@dougbu For AspLabs you should be able to update the version here -> https://github.com/dotnet/arcade/blob/f7136626d0109856df867481219eb7366951985d/eng/common/templates/job/job.yml#L36  

Also update on the linux musl leg, I tried to run the build with updated verbosity, that did not work. We are waiting on SBOM folks for that.

**Comment by RussKie** [comment]
@epananth @dreddy-work please feel to merge, if these look correct.
* https://github.com/dotnet/winforms/pull/6759 -> https://dev.azure.com/dnceng/internal/_build/results?buildId=1631751&view=results
* https://github.com/dotnet/windowsdesktop/pull/2651 -> https://dev.azure.com/dnceng/internal/_build/results?buildId=1631811&view=results

Release/6.0 branches (neither public nor internal) for both repos don't appear to have generate-sbom.yml. Is this something coming?

**Comment by epananth** [comment]
@RussKie  Thank you! 
We already have backported this to arcade release/6.0 https://github.com/dotnet/arcade/pull/8479, you should have an arcade update for your repo for that..

**Comment by joeloff** [comment]
@mmitche I talked to to @epananth The SDK SBOM is 1.4 million lines. It crashed VS, VSCode. I finally managed to open it in Notepad++, but reducing size of the files should be a priority next if possible.

**Comment by RussKie** [comment]
In dotnet/winforms it's a measly 114K lines... There are lot of entries for non-prod artifacts (e.g., tests), which could probably be ignored.

**Comment by zivkan** [comment]
NuGet's PR is merged now, and new builds will generate the sbom file & build artifact. Our next insertion will be next week.

**Comment by RussKie** [comment]
https://github.com/dotnet/winforms/pull/6759 is merged. 
https://github.com/dotnet/windowsdesktop/pull/2651 is failing to generate SBOM on ["Prepare for publish" leg](https://github.com/dotnet/windowsdesktop/pull/2651#issuecomment-1051005993). See https://dev.azure.com/dnceng/internal/_build/results?buildId=1633615&view=logs&j=5ab303af-16db-5f58-82d4-945dcabe3bb5&t=24962100-60d4-5768-b736-1d1c025ebd15. I'd appreciate guidance.

**Comment by epananth** [comment]
@RussKie looking into this

**Comment by MiYanni** [comment]
@epananth The `dotnet-project-system` has now been verified to have SBOM. The insertion PR hasn't merged yet, but the SBOM check has passed.
https://devdiv.visualstudio.com/DevDiv/_git/VS/pullrequest/385416

**Comment by epananth** [comment]
> @epananth The `dotnet-project-system` has now been verified to have SBOM. The insertion PR hasn't merged yet, but the SBOM check has passed. https://devdiv.visualstudio.com/DevDiv/_git/VS/pullrequest/385416

Thanks @MiYanni. I updated the list

**Comment by hoyosjs** [comment]
Forgot to update - dotnet-symuploader is good. I am not sure who deals with internal-components atm

**Comment by epananth** [comment]
> Forgot to update - dotnet-symuploader is good. I am not sure who deals with internal-components atm

Thanks @hoyosjs

**Comment by epananth** [comment]
Calling this done and closing the issue.

**Comment by Tratcher** [comment]
Completed for aspnetkatana.
https://dev.azure.com/dnceng/internal/_build/results?buildId=1700756&view=logs&j=0bc77094-9fcd-5c38-f6e4-27d2ae131589&t=f5edc144-6ff8-5609-0882-2ee7397b69c1

**Comment by StephenMolloy** [comment]
ASP.Net Classic nuget package pipelines have been updated.
